pkg/operator/targetconfigcontroller/targetconfigcontroller.go

```diff
@@ -148,6 +148,16 @@ func isRequiredConfigPresent(config []byte) error {
 		if configValString, ok := configVal.(string); ok && len(configValString) == 0 {
 			return fmt.Errorf("%v empty in config", strings.Join(requiredPath, "."))
 		}
+
+		if requiredPath[0] == "etcd-servers" {
+			configValSlice, ok := configVal.([]interface{})
+			if !ok {
+				return fmt.Errorf("%v is not a slice", strings.Join(requiredPath, "."))
+			}
+			if len(configValSlice) < 3 {
```

Comment:
This needs to be gated for SNO as I assume it would only get 2 endpoints. Also, shouldn't we aim for 4 endpoints since we include localhost?

Reply:
yes, adding an SNO gate too.

>shouldn't we aim for 4 endpoints since we include localhost?

Three endpoints - localhost, local IP and etcd IP on any other master - should be sufficient to prevent hangup for 10 mins.
We may never have all 4 endpoints on HA with assisted installer - one of masters is used as a bootstrap, so we shouldn't rely on having all 3 masters available during bootstrap.

Reply:
I haven’t read the code, but I’m wondering if we only want to stop new revisions during the installation time. If not, what happens when a customer takes a master node out of the pool for service? Does that mean the kas-o will stop installing new revisions? If so, that’s bad, right?

Reply:
Also, recently there was a change to the RevisionController (library-go) which permits passing a precondition function. New revisions will be allowed by the controller once the precondition fulfils. Maybe you could use it.

Reply:
>If not, what happens when a customer takes a master node out of the pool for service?

Multiple scenarios here:
* The node may be physically down for maintenance, but it would still be listed as an endpoint. 

* Also, while we require three endpoints - its in fact two distinct etcd servers (local IP and localhost are considered single etcd server but two distinct endpoints). So this code won't prevent one master from being torn down during CPMS scale up / down

* However there is an etcd restore procedure where the cluster is reduced to a single node to restore to backup and later being scale up to 3. We specifically don't want rolling out new revisions until all three masters are available to ensure all three etcd servers are in sync.

TODO: run e2e to do cpms scale up / etcd disruption to ensure that it works as expected

Reply:
>New revisions will be allowed by the controller once the precondition fulfils

`WithRevisionControllerPrecondition` might be an alternative to this solution, yes

Reply:
> WithRevisionControllerPrecondition might be an alternative to this solution, yes

Not a blocking comment just wanted to let you know the mechanism exist so that you may consider using it.

So, when an etcd endpoint is removed from the list?

Reply:
Also, should we add a similar mechanism for the aggregated API servers ?

Reply:
>just wanted to let you know the mechanism exist so that you may consider using it

Right, I think current one is more suited for this situation though - we already have a config validation function so its only natural to extend it. This way we prevent the config from being generated not merely rolled out.

However if this method won't work in some scenarios `WithRevisionControllerPrecondition` would be a good candidate to try.

>should we add a similar mechanism for the aggregated API servers 

Its not necessary imo, the issue is happening early on bootstrap when openshift-apiserver is not even created. 

If kube-apiserver can reliably write to etcd we'll notice slow etcd rollout via "static pod didn't rollout in 3 mins" test - and a similar solution can be applied to openshift-apiservers